test/e2e/reconciler/workspaceindex/controller_test.go

```diff
@@ -0,0 +1,452 @@
+					return
+				}
+				if err := resp.Body.Close(); err != nil {
+					t.Errorf("failed to close body: %v", err)
+					return
+				}
+				t.Log(string(body))
+
+				for orgName, workspaces := range observed {
+					for workspaceName, shardName := range workspaces {
+						logicalCluster := orgName + "_" + workspaceName
+						shard, err := resolve(logicalCluster, "")
+						if err != nil {
+							t.Errorf("%s/%s: expected no error but got one: %v", orgName, workspaceName, err)
+							return
+						}
+						if shard != shardName {
+							t.Errorf("%s/%s: expected %s on %s, got %s", orgName, workspaceName, logicalCluster, shardName, shard)
+							return
+						}
+					}
+				}
+			},
+		},
+	}
+	const (
+		// serverNameMain will hold the root workspace
+		serverNameMain = "main"
+		// serverName* will hold org workspaces
+		serverNameEast    = "east"
+		serverNameCentral = "central"
+		serverNameWest    = "west"
+		// orgName* are the organizations
+		orgNameAcme    = "acme"
+		orgNameInitech = "initech"
+		orgNameGlobex  = "globex"
+	)
+	for i := range testCases {
+		testCase := testCases[i]
+		framework.Run(t, testCase.name, func(t framework.TestingTInterface, servers map[string]framework.RunningServer, artifactDir, dataDir string) {
+			ctx := context.Background()
+			if deadline, ok := t.Deadline(); ok {
+				withDeadline, cancel := context.WithDeadline(ctx, deadline)
+				t.Cleanup(cancel)
+				ctx = withDeadline
+			}
+			if len(servers) != 4 {
+				t.Errorf("incorrect number of servers: %d", len(servers))
+				return
+			}
+			mainServer, err := resolveRunningServer(ctx, t, servers[serverNameMain], "")
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			// in the root workspace, set up:
+			// - WorkspaceShard objects with the credentials for our servers
+			// - Workspace objects for our organizations
+			// in the "admin" workspace on each shard, set up:
+			// - WorkspaceShard object for the scheduler (this will be removed when we have a cross-shard client)
+			// - Workspace objects (these are end-user-facing)
+			serverNames := []string{serverNameEast, serverNameCentral, serverNameWest}
+			orgs := []string{orgNameAcme, orgNameInitech, orgNameGlobex}
+			workspaces := []string{"accounting", "engineering", "executive"}
+			initErrorChan := make(chan error, len(serverNames)+len(orgs))
+			wg := sync.WaitGroup{}
+			wg.Add(len(serverNames))
+			initStart := time.Now()
+			t.Log("bootstrapping WorkspaceShards in root logical cluster")
+			for _, serverName := range serverNames {
+				go func(serverName string) {
+					defer wg.Done()
+
+					rawCfg, err := servers[serverName].RawConfig()
+					if err != nil {
+						initErrorChan <- fmt.Errorf("failed to resolve raw credentials: %w", err)
+						return
+					}
+					if err := initializeShard(ctx, t, mainServer, serverName, rawCfg); err != nil {
+						initErrorChan <- err
+						return
+					}
+				}(serverName)
+			}
+			close(initErrorChan)
```

Comment:
Could this result in panic (send on closed channel) in the goroutines above? Does it need to be moved after wg.Wait()?

Reply:
Good catch1